velox/dwio/common/IntDecoder.h

```diff
@@ -437,24 +436,28 @@ inline T IntDecoder<isSigned>::readInt() {
   if (useVInts_) {
     return readVInt<T>();
   }
+
   if (bigEndian_) {
     return readLittleEndianFromBigEndian<T>();
-  } else {
-    if constexpr (std::is_same_v<T, int128_t>) {
-      if (numBytes_ == 12) {
-        VELOX_DCHECK(!useVInts_, "Int96 should not be VInt encoded.");
-        return readInt96();
-      }
-      VELOX_NYI();
+  }
+
+  if constexpr (std::is_same_v<T, int128_t>) {
+    // Int64 or Int96 in Parquet
+    if (numBytes_ == 8 || numBytes_ == 12) {
+      return readSignedIntAs128();
     }
-    return readLongLE();
+    VELOX_NYI();
   }
+
+  return readLongLE();
 }
 
 template <bool isSigned>
-inline int128_t IntDecoder<isSigned>::readInt96() {
+inline int128_t IntDecoder<isSigned>::readSignedIntAs128() {
+  VELOX_DCHECK(!useVInts_, "Should not be VInt encoded.");
+
   int128_t result = 0;
-  for (int i = 0; i < 12; ++i) {
+  for (int i = 0; i < numBytes_; ++i) {
```

Comment:
You instead call readLongLE() when numBytes_ == 8, that will work fine for negative values as well